omniledger/service/viewchange.go

```diff
@@ -0,0 +1,371 @@
+	for i := range proof {
+		if proof[i].SignerID.Equal(s.ServerIdentity().ID) && len(proof[i].Signature) == 0 {
+			proof[i].Sign(s.getPrivateKey())
+		}
+	}
+
+	sb := s.db().GetByID(proof[0].View.ID)
+	req := viewchange.NewViewReq{
+		Roster: *rotateRoster(sb.Roster, proof[0].View.LeaderIndex),
+		Proof:  proof,
+	}
+
+	go func() {
+		// This go-routine eventually exists because both cosi and
+		// block creation have a timeout.
+		sig, err := s.startViewChangeCosi(req)
+		if err != nil {
+			log.Error(s.ServerIdentity(), err)
+			return
+		}
+		if len(sig) == 0 {
+			log.Error(s.ServerIdentity(), "empty viewchange cosi signature")
+			return
+		}
+		if err := s.createViewChangeBlock(req, sig); err != nil {
+			log.Error(s.ServerIdentity(), err)
+		}
+	}()
+}
+
+func (s *Service) computeInitialDuration(scID skipchain.SkipBlockID) time.Duration {
+	interval, err := s.LoadBlockInterval(scID)
+	if err != nil {
+		panic("we should only computer the interval on a known block," +
+			" otherwise it is a progarmmer error: " + err.Error())
+	}
+	return rotationWindow * interval
+}
+
+func (s *Service) getFaultThreshold(sbID skipchain.SkipBlockID) int {
+	sb := s.db().GetByID(sbID)
+	return len(sb.Roster.List) / 3
+}
+
+// handleViewChangeReq should be registered as a handler for viewchange.InitReq
+// messages.
+func (s *Service) handleViewChangeReq(env *network.Envelope) {
+	// Parse message.
+	req, ok := env.Msg.(*viewchange.InitReq)
+	if !ok {
+		log.Error(s.ServerIdentity(), "failed to cast to viewchange.ViewChangeReq")
+		return
+	}
+	// Should not be sending to ourselves.
+	if req.SignerID.Equal(s.ServerIdentity().ID) {
+		log.Error(s.ServerIdentity(), "should not send to ourselve")
+		return
+	}
+
+	// Check that the genesis exists and the view is valid.
+	_, err := s.db().GetLatestByID(req.View.Gen)
+	if err != nil {
+		log.Error(s.ServerIdentity(), err)
+		return
+	}
+	reqLatest := s.db().GetByID(req.View.ID)
+	if reqLatest == nil {
+		// NOTE: If we don't know about the this view, it might be that
+		// we are not up-do-date, which should not happen because the
+		// delay for triggering view-change should be longer than the
+		// time it takes to create and propagate a new block. Hence,
+		// somebody is sending bogus views.
+		log.Error(s.ServerIdentity(), "we do not know this view")
+		return
+	}
+	actualLatest, err := s.db().GetLatest(reqLatest)
+	if err != nil || !actualLatest.Hash.Equal(reqLatest.Hash) {
+		log.Error(s.ServerIdentity(), "viewchange should not happen for earlier blocks", err)
+		return
+	}
+
+	// Check signature.
+	_, signerSID := actualLatest.Roster.Search(req.SignerID)
+	if signerSID == nil {
+		log.Error(s.ServerIdentity(), "signer does not exist")
+		return
+	}
+	if err := schnorr.Verify(cothority.Suite, signerSID.Public, req.Hash(), req.Signature); err != nil {
+		log.Error(s.ServerIdentity(), err)
+		return
+	}
+
+	// Store it in our log.
+	s.viewChangeMan.addReq(*req)
```

Comment:
I generally avoid using pointers because when something cannot be nil, then you eliminate one source of error - nil pointer de-reference. Further, I think the extra memory you push on top of the stack isn't a concern, it always gets poped when the function returns. For these reasons, I almost always write functions that take pointers unless the struct is huge or it needs to be modified. The reason I have to de-reference here is because `protobuf.Decode` returns a pointer.

Reply:
ok got it. So paraphrasing: in the absence of a compelling reason (huge struct or perf measurement) you opt for NOT passing pointers because then you don't need to check for nil in the callee. I can live with that. I might forget in my own code, feel free to call me on it.